File: Leetcode/1387. Sort Integers by The Power Value/solution.cpp
class Solution {
public:
    unordered_map<int, int> mem;

    int compute(int n) {
        if (mem[n] || n == 1) {
            return mem[n];
        }

        int count = 0;
        int power = n;

        while (power != 1) {
            if (power % 2) {
                power = power * 3 + 1;
            }
            else {
                power = power / 2;
            }
            count++;
        }

        mem[n] = count;

        return mem[n];
    }

    int getKth(int lo, int hi, int k) {
        vector<pair<int, int>> arr;

        for (int i = lo; i <= hi; i++) {
            arr.push_back({ compute(i), i });
        }

        sort(arr.begin(), arr.end());
        return arr[k-1].second;
    }
};

File: Leetcode/1387. Sort Integers by The Power Value/solution.py
class Solution:
    mem = {1: 1}

    def compute(self, n):
        if n == 1 or n in self.mem:
            return self.mem[n]
        if n % 2:
            self.mem[n] = self.compute(n * 3 + 1) + 1
            return self.mem[n]
        else:
            self.mem[n] = self.compute(n / 2) + 1
            return self.mem[n]

    def getKth(self, lo: int, hi: int, k: int) -> int:

        for i in range(lo, hi+1):
            self.compute(i)

        return sorted((self.mem[i], i) for i in range(lo, hi+1))[k - 1][1]
